File: apps/mobile/modules/expo-file-hash/android/src/main/java/expo/modules/filehash/ExpoFileHashModule.kt
package expo.modules.filehash

import android.net.Uri
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.FileInputStream
import java.io.InputStream
import java.security.MessageDigest

class ExpoFileHashModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoFileHash")

    AsyncFunction("sha256") { uri: String ->
      val context = appContext.reactContext ?: throw FileHashException(uri)
      val parsed = Uri.parse(uri)
      val input: InputStream = when (parsed.scheme) {
        "content" -> context.contentResolver.openInputStream(parsed)
          ?: throw FileHashException(uri)
        "file" -> FileInputStream(parsed.path ?: throw FileHashException(uri))
        else -> FileInputStream(uri)
      }

      input.use { stream ->
        val digest = MessageDigest.getInstance("SHA-256")
        val buffer = ByteArray(1024 * 1024)
        while (true) {
          val count = stream.read(buffer)
          if (count <= 0) break
          digest.update(buffer, 0, count)
        }
        digest.digest().joinToString("") { "%02x".format(it) }
      }
    }
  }
}

private class FileHashException(uri: String) :
  CodedException("ERR_FILE_HASH", "Unable to calculate SHA-256 for file: $uri", null)

File: apps/mobile/modules/expo-file-hash/ios/ExpoFileHashModule.swift
import CryptoKit
import ExpoModulesCore

public class ExpoFileHashModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoFileHash")

    AsyncFunction("sha256") { (uri: String) throws -> String in
      let url: URL
      if let parsed = URL(string: uri), parsed.isFileURL {
        url = parsed
      } else {
        url = URL(fileURLWithPath: uri)
      }

      guard let stream = InputStream(url: url) else {
        throw FileHashException(uri)
      }

      stream.open()
      defer { stream.close() }

      var hasher = SHA256()
      let bufferSize = 1024 * 1024
      let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
      defer { buffer.deallocate() }

      while stream.hasBytesAvailable {
        let count = stream.read(buffer, maxLength: bufferSize)
        if count < 0 {
          throw FileHashException(uri)
        }
        if count == 0 {
          break
        }
        hasher.update(data: Data(bytes: buffer, count: count))
      }

      return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }
  }
}

private class FileHashException: GenericException<String> {
  override var reason: String {
    "Unable to calculate SHA-256 for file: \(param)"
  }
}
